streaming/src/decrypt_reader.rs

```diff
@@ -0,0 +1,201 @@
+    aad: Vec<u8>,
+    state: State,
+}
+
+impl DecryptReader {
+    pub fn new(
+        wrapped: crate::WrappedStreamingAead,
+        reader: Box<dyn io::Read>,
+        aad: &[u8],
+    ) -> Self {
+        Self {
+            wrapped,
+            aad: aad.to_vec(),
+            state: State::Pending(reader),
+        }
+    }
+}
+
+impl io::Read for DecryptReader {
+    fn read(&mut self, p: &mut [u8]) -> io::Result<usize> {
+        match &mut self.state {
+            State::Found(reader) => return reader.read(p),
+            State::Failed => {
+                return Err(io::Error::new(
+                    io::ErrorKind::InvalidInput,
+                    "no matching key found for the ciphertext in the stream",
+                ))
+            }
+            State::Pending(_) => {}
+        };
+        // Move the underlying raw reader out of self and into a `SharedCopyReader`
+        let state = std::mem::replace(&mut self.state, State::Failed);
+        let raw_reader = match state {
+            State::Pending(reader) => reader,
+            _ => unreachable!(),
+        };
+        let mut copy_reader = SharedCopyReader::new(raw_reader);
+
+        // find proper key to decrypt ciphertext
+        let entries = self.wrapped.ps.raw_entries();
+        for e in &entries {
+            let sa = match &e.primitive {
+                tink::Primitive::StreamingAead(p) => p,
+                _ => continue,
+            };
+
+            // Attempt a decrypting-read from the ciphertext reader `cr`, but also keep a copy of
+            // the read data into a buffer so that it can be re-scanned with a different key if
+            // decryption fails.
+            let mut r = match sa.new_decrypting_reader(Box::new(copy_reader.clone()), &self.aad) {
+                Ok(r) => r,
+                Err(_) => {
+                    copy_reader.rewind();
+                    continue;
+                }
+            };
+            let n = match r.read(p) {
+                Ok(n) => n,
+                Err(_) => {
+                    // The read attempt will have consumed some of the underlying reader, but
+                    // there is a copy of the data that has been read. Ensure that this already-read
+                    // data is re-used next time around.
+                    copy_reader.rewind();
+                    continue;
+                }
+            };
+
+            // Reading has succeeded, so use this particular key from now on and no longer need
+            // to store copies of read data.
+            copy_reader.stop_copying();
+            self.state = State::Found(r);
+            return Ok(n);
+        }
+        Err(io::Error::new(
+            io::ErrorKind::InvalidInput,
+            "no matching key found for the ciphertext in the stream",
+        ))
+    }
+}
+
+/// Wrapper around an [`io::Read`] trait object that stores a copy of all of the data
+/// read from the underlying object.
+struct CopyReader {
```

Comment:
Yeah I didn't mean peekable per se, but implementing a `PeekableRead` or similar. Similar to your `CopyReader`, but more transparent as it would not need to be rewound, it would just transparently collect the peeked data in an internal buffer when calling a `peek` method, and then transparently return it as prefix during the next call to `read`. But if you prefer the explicit `start` / `rewind`, yours looks fine.